openapi: Document /realm endpoint.

clarammdantas opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR adds openapi documentation to `/realm` endpoint.

clarammdantas committed:
openapi: Document /realm endpoint on zulip.yaml.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -2924,6 +2924,323 @@ paths:
+          An integer that defines who can create streams.
+          1 = Admins and member can create streams.
+          2 = Only admins can create streams.
+          3 = Admins and full members can create streams.
```
Let's make these groups bulleted lists.

clarammdantas replied:
Can I still let the values of each option? Personally speaking, I find it very handy when the API documentation have the values for an argument such as this one. I've committed as bulleted lists, with the values, but if it's better off, I change it.

timabbott replied:
Yeah, just to be clear, I want these details here -- just a bullet before the number `1` (You'll see when you render this; without the bullets I think everything will flow to 1 line with that markdown input)

timabbott commented:
Oh, one other comment: Please don't spend much (any?) time documenting the response.  We're going to want to change this endpoint's response to use this pattern: https://chat.zulip.org/#narrow/stream/3-backend/topic/update-subscription-properties/near/901757

clarammdantas commented:
> Oh, one other comment: Please don't spend much (any?) time documenting the response. We're going to want to change this endpoint's response to use this pattern: https://chat.zulip.org/#narrow/stream/3-backend/topic/update-subscription-properties/near/901757

oops sorry, I just documented it.

timabbott commented:
@orientor it'd be great if you can do the first review on this.

orientor commented on zerver/openapi/zulip.yaml:
```diff
@@ -3018,6 +3018,428 @@ paths:
+        example: false
+        required: false
+      - name: invite_by_admins_only
+        in: quey
```
Typo here. It should be `query` .

orientor commented on zerver/openapi/zulip.yaml:
```diff
@@ -3018,6 +3018,428 @@ paths:
+        example: 3
+        required: false
+      - name: authentication_methods
+        type: object
```
The type should be inside `schema`. Also all `array` and `object` parameters should have an example.

orientor reviewed (commented):
@clarammdantas thanks for working on this! I think the next steps would be to create a `update-realm.md` in `templates/zerver/api/`. You can use the `dev-fetch-api-key.md` as template with few changes. Then this `\realm` endpoint should be removed from the exclude-endpoint list in `zerver/tests/test_openapi.py`. Also as the response is not to be documented you can add `/realm:patch` to `EXCLUDE_ENDPOINTS` in `zerver/openapi/openapi.py`. Also finally add the new .md to `templates/zerver/help/include/rest-endpoints.md`.

clarammdantas committed:
test_openapi: Remove /realm from pending_endpoints.

clarammdantas committed:
templates: Add template that documents /realm endpoint.

clarammdantas committed:
python_example: Add python example for /realm:patch.

clarammdantas committed:
test-api: Add owner_client.

There are a few organization settings that only organization
owners are allowed to change. After documenting the /realm:patch
it was needed to add a owner_client to the test-api, making it
possible to test the /realm:patch endpoint.

clarammdantas committed:
openapi: Skip validation of undocumented responses.

The /realm endpoint's response will change soon, so we didn't put
much effort into documenting it. That are a lot of undocumented
keys on the /realm response, so instead of writing them all in
the EXCLUDE_PROPERTIES, we created another structure,
EXCLUDE_DOCUMENTED_ENDPOINTS, to those endpoints that should be
skipped because of poor response documentation.

Adding the endpoint to EXCLUDE_ENDPOINTS didn't work because there
is a documentation for /realm:patch, but the response is missing
most of the keys documentation. Also, if the endpoint is in this list,
it breaks the tests on test_openapi because we registered /realm:patch
and the validate_against_openapi_schema return False when it skips
endpoints on this list.

For more information about the plans to change this endpoint's response,
see discussion on #15417.

clarammdantas commented:
@orientor thanks for your review, it was helpful and complementary to the documentation on how to write API documentation. This PR is now ready for review.

I did a different thing from your suggestion to add `/realm:patch` to the `EXCLUDE_ENPOINTS` list. The thing is that in order to render the `update-organization.md` file, I needed to register the `/realm:patch` endpoint. It kept breaking because of the `validate_against_openapi_schema` function return `False` for skipped endpoints.

I didn't lose that much time documenting the response as Tim suggested and my workaround to not document all the keys in the response was to add a dictionary like the `EXLUDE_PROPERTIES` but instead, hold all the documented keys, so I could ignore everything that wasn't there for `/realm:patch`. There are a lot of places in backend tests that call for `client_patch('/json/realm')` with lots of different keys, which would break my code when validating against the schema.

Adding the undocumented keys to `EXLUDE_PROPERTIES` would work, but I would lose a lot of time writing down all the arguments on `update_realm`.

One more thing, if it isn't much trouble for you, please comment if there is anything on my commits writing or structure that I could improve to make it easier to review. I would really appreciate that! :)

orientor commented on zerver/openapi/openapi.py:
```diff
@@ -275,6 +288,8 @@ def validate_against_openapi_schema(content: Dict[str, Any], endpoint: str,
     # In a single response schema we do not have two keys with the same name.
     # Hence exclusion list is declared globally
     exclusion_list = (EXCLUDE_PROPERTIES.get(endpoint, {}).get(method.lower(), {}).get(response, []))
+
```
Can you create another array `EXCLUDE_DOCUMENTED_ENDPOINTS` and then add code similar to this right after line 283?(I can't seem to link it to review)
```
    if endpoint + ':' + method in EXCLUDE_DOCUMENTED_ENDPOINTS:
        return True
```
I think this will be a better method as in the future we may have arrays or objects in `DOCUMENTED_RESPONSE_KEYS` which may lead to some problems. Also an endpoint will be in `DOCUMENTED_RESPONSE_KEYS` only if it's response is mostly undocumented. In that case validation will not be too meaningful and we would be better off skipping it and returning True.

orientor commented on zerver/openapi/zulip.yaml:
```diff
@@ -3095,6 +3095,458 @@ paths:
+        in: query
+        description: |
+          An integer that indicates the day that the weekly digest emails will
+          be sent to inactive users.
```
The bulleted lists are all coming in one line in the browser. You can have a look at `update-stream` for the bulleted lists in parameters.

orientor reviewed (commented):
@clarammdantas left a few comments. Other than those everything else LGTM. Thanks for working on this! Regarding the commits, they are very clear and modularized for reviewing so I don't think there's anything I can add.

clarammdantas commented on zerver/openapi/python_examples.py:
```diff
@@ -1215,8 +1232,9 @@ def test_queues(client: Client) -> None:
-def test_server_organizations(client: Client) -> None:
+def test_server_organizations(client: Client, owner_client: Client) -> None:
 
+    update_realm(client)
```
That's because, in the python example, we are only trying to change the `digest_weekday` and `signup_notifications_stream_id` both can be changed by an admin. But in the `curl_example,` it tries to update all the attributes that were defined on `zulip.yaml` for the `/realm:patch` and a couple of them can only be changed by an organization owner.
I let the `owner_client` do the changes on `python_examples` just in case someone wanted a different example, it wouldn't break with `Desdemona` for sure but it can break with `Iago`.

clarammdantas commented:
Ok, so it's not possible to name the `.md` file as `update-realm` I tried to do a workaround, but it failed anyway. The lint doesn't allow to have realm references in the `rest-endpoints.md`. Unless we disable the lint for that line: `* [Update organization](/api/update-organization)` but I don't know if that's a good practice and something people do in this project.

orientor commented:
Ok so we should have something like this : `[Update orgainzation](api/update-realm)`. Now for this line we will definitely need the linter check as this is the most likely place where someone may turn organization to realm. And we also need /update-realm as the naming convention requires it. Furthermore the `update-realm` link can not be considered to be user-facing as it is just a link. So the linter should not catch it.

I think the linter regex should be changed to `(^|\s)[rR]ealm[s]?\b` This matches the beginning of the word with either the beginning of a line or a space which should ideally be how we match a word in a line. This works for our case and also passes the linter tests.

@timabbott can you take a look?

timabbott commented:
Yeah, that regex change sounds like a good solution to me.

timabbott commented:
@orientor feel free to submit that regex change as a separate PR (or @clarammdantas feel free to add it to the start of your branch).  We should never allow linters to make us do the wrong thing; we can always add exclude rules or improve the linters; they are a tool to help write good code, not something that can help decide what's correct :)

timabbott commented on zerver/openapi/test_curl_examples.py:
```diff
@@ -27,6 +27,11 @@ def test_generated_curl_examples_for_success(client: Client) -> None:
     # very explicit.
     for file_name in sorted(glob.glob("templates/zerver/api/*.md")):
+        if 'update-realm' in file_name:
+            authentication_line = f"{owner_client.email}:{owner_client.api_key}"
```
Is there a way we can pass a flag to control which client to use, rather than having the endpoint be the conditional here?

clarammdantas replied:
@timabbott I don't know if that was the expected solution, but I tried to add a flag on the `.md` file so every file can specify which kind of user to use. So, you can have `use_admin_client`, `use_owner_client`, `use_guest_client` and so on.

clarammdantas committed:
custom_check: Add new regex to check usage of the word realm.

We refer to "realms" as "organizations" in user-facing docs, but
the linter shouldn't catch the word "realm" if it's on a URL, as
in /api/update-realm at rest-endpoints.md file.

clarammdantas commented:
I've updated this PR. It's ready for review.

aero31aero commented on zerver/openapi/openapi.py:
```diff
@@ -95,6 +95,16 @@
+# Set of documented endpoints that doesn't have a documented response or have an
+# incomplete documented response.
+EXCLUDE_DOCUMENTED_ENDPOINTS = {
+    # This endpoint have a lot of keys and the response is not fully documented
```
`s/have/has`.

Also, just below this we have `EXCLUDE_ENDPOINTS` list which this implementation is a clone of. We can just add `/realm:patch` to `EXCLUDE_ENDPOINTS`.

orientor replied:
`validate_against_openapi_schema` returns true if validation happens else false. For the second array it consists of undocumented endpoint and returns false. But for the documented endpoints we want true to make sure validation is happening. Hence another array for endpoints with only response documentation remaining so that the function will return true for them.

timabbott commented:
@clarammdantas would you be up for rebasing this?  If we don't hear from you in the next couple days, I'll assume someone else should pick it up.

@orientor @MSurfer20 this is a nice chunk of work that it looks like @orientor was happy with as of last summer, but we didn't get in.  It'd be great to incorporate and close out.

clarammdantas commented:
> @clarammdantas would you be up for rebasing this? If we don't hear from you in the next couple days, I'll assume someone else should pick it up.
> 
> @orientor @MSurfer20 this is a nice chunk of work that it looks like @orientor was happy with as of last summer, but we didn't get in. It'd be great to incorporate and close out.

@timabbott  I'll be happy to do it! Tomorrow I'll rebase it and check if there is something else left to be done

timabbott commented:
Cool.  Make sure you read the chat.zulip.org thread on the black migration
since it provides an easy way to rebase past when we reformatted the
codebase.

-Tim Abbott (mobile)

On Mon, May 24, 2021, 19:36 Clara Moraes Dantas ***@***.***>
wrote:

> @clarammdantas <https://github.com/clarammdantas> would you be up for
> rebasing this? If we don't hear from you in the next couple days, I'll
> assume someone else should pick it up.
>
> @orientor <https://github.com/orientor> @MSurfer20
> <https://github.com/MSurfer20> this is a nice chunk of work that it looks
> like @orientor <https://github.com/orientor> was happy with as of last
> summer, but we didn't get in. It'd be great to incorporate and close out.
>
> @timabbott <https://github.com/timabbott> I'll be happy to do it!
> Tomorrow I'll rebase it and check if there is something else left to be done
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/15417#issuecomment-847485490>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAU6NWVVA74OFDLB6RIUSQDTPMEKTANCNFSM4OAFSMCA>
> .
>

clarammdantas commented:
Just an update on this: I did the rebase, and I've followed the tutorial on general to migrate to Black. But the tests were breaking, so today I'm going to try to fix them so I can push it here.

timabbott commented:
Awesome!
